File: android/data/services/gemstone/src/main/kotlin/com/gemwallet/android/data/services/gemstone/stores/NftStore.kt
package com.gemwallet.android.data.services.gemstone.stores

import com.gemwallet.android.data.services.store.database.NftDao
import com.gemwallet.android.data.services.store.database.entities.DbNFTAssociation
import com.gemwallet.android.data.services.store.database.entities.toNFTAsset
import com.gemwallet.android.data.services.store.database.entities.toNFTCollection
import com.gemwallet.android.data.services.store.database.entities.toRecord
import com.gemwallet.android.ext.toGem
import com.gemwallet.android.ext.toNftAssetId
import com.gemwallet.android.ext.toPrimitives
import com.wallet.core.primitives.NFTAssetData
import com.wallet.core.primitives.NFTAssetId
import kotlinx.coroutines.flow.first
import uniffi.gemstone.GemNftStore

class GemstoneNftStore(private val nftDao: NftDao) : GemNftStore {

    override suspend fun saveNfts(walletId: String, data: List<uniffi.gemstone.NftData>) {
        val nftData = data.map { it.toPrimitives() }
        val assets = nftData.flatMap { it.assets }.map { it.toRecord() }
        nftDao.updateNft(
            walletId = walletId,
            collections = nftData.map { it.collection.toRecord() },
            assets = assets,
            associations = assets.map { DbNFTAssociation(walletId = walletId, assetId = it.id) },
        )
    }

    override suspend fun getAssetData(assetId: String): uniffi.gemstone.NftAssetData? {
        val id = NFTAssetId(assetId)
        val asset = nftDao.getAsset(id).first() ?: return null
        val collection = nftDao.getCollection(asset.collectionId).first() ?: return null
        return NFTAssetData(collection = collection.toNFTCollection(), asset = asset.toNFTAsset()).toGem()
    }

    override suspend fun saveAsset(data: uniffi.gemstone.NftAssetData) {
        val assetData = data.toPrimitives()
        nftDao.add(collection = assetData.collection.toRecord(), asset = assetData.asset.toRecord())
    }
}

File: ios/Packages/GemstoneServices/Sources/Stores/NftStore.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import struct Gemstone.NftAssetData
import typealias Gemstone.NftAssetId
import struct Gemstone.NftData
import protocol Gemstone.GemNftStore
import GemstonePrimitives
import Primitives
import Store

public final class GemstoneNftStore: GemNftStore, @unchecked Sendable {
    private let store: NftStore

    public init(store: NftStore) {
        self.store = store
    }

    public func saveNfts(walletId: String, data: [Gemstone.NftData]) async throws {
        try store.save(data.map { $0.map() }, for: WalletId.from(id: walletId))
    }

    public func getAssetData(assetId: Gemstone.NftAssetId) async throws -> Gemstone.NftAssetData? {
        guard let asset = try store.getAsset(assetId: NFTAssetId.from(id: assetId)),
              let collection = try store.getCollection(collectionId: asset.collectionId)
        else {
            return nil
        }
        return NFTAssetData(collection: collection, asset: asset).map()
    }

    public func saveAsset(data: Gemstone.NftAssetData) async throws {
        let data = data.map()
        try store.add(asset: data.asset, collection: data.collection)
    }
}
